### PR TITLE
Remove Helper Method Frames (HMF) from Reflection

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Convert `RuntimeMethodHandle.GetDeclaringType()` to managed and QCall.
Convert `RuntimeTypeHandle.GetElementType()` to managed.
Correct nullability in signature and usage.
Convert `RuntimeTypeHandle.GetDeclaringType()` to managed and QCalls.
Remove `GetRuntimeTypeHelper()`.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
@EgorBot -intel -arm64 -perf
```
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Bench>(args: args);

public class Bench
{
    struct S;

    [Benchmark]
    public Type Nested() => typeof(S).DeclaringType;

    [Benchmark]
    public Type NotNested() => typeof(Bench).DeclaringType;

    [Benchmark]
    public Type RefElementType() => typeof(Bench[]).GetElementType();

    [Benchmark]
    public Type ValueElementType() => typeof(S[]).GetElementType();
}
```

**Comment by EgorBo** [comment]
@AaronRobinsonMSFT the bot complained that `-profiler` flag should be used when there are not more than 3 [Benchmark] attributes. I think I should improve error reporting 🙂

**Comment by EgorBo** [comment]
@EgorBot -x64 -arm64

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    struct S;

    [Benchmark]
    public Type Nested() => typeof(S).DeclaringType;

    [Benchmark]
    public Type NotNested() => typeof(Bench).DeclaringType;

    [Benchmark]
    public Type RefElementType() => typeof(Bench[]).GetElementType();

    [Benchmark]
    public Type ValueElementType() => typeof(S[]).GetElementType();
}
```

**Comment by AaronRobinsonMSFT** [comment]
@EgorBot -x64 -arm64 -profiler

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    struct S;

    [Benchmark]
    public Type RefElementType() => typeof(Bench[]).GetElementType();

    [Benchmark]
    public Type ValueElementType() => typeof(S[]).GetElementType();
}
```

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 1]
```diff
@@ -928,7 +967,25 @@ internal static string ConstructInstantiation(IRuntimeMethodInfo method, TypeNam
...
+        {
+            Debug.Assert(!method.IsNullHandle());
+            MethodTable* pMT = GetMethodTable(method);
+            if (pMT->IsArray)
+            {
+                IntPtr handle = GetHandleForArray(pMT);
+                return RuntimeTypeHandle.GetRuntimeTypeFromHandle(handle);
+            }
+            else
+            {
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:985]:
```suggestion
```
I do not think that this is needed. We got rid of the sharing of methodtable for arrays.

File: src/coreclr/vm/runtimehandles.cpp [hunk 2]
```diff
@@ -1007,117 +981,84 @@ FCIMPL1(ReflectMethodObject*, RuntimeTypeHandle::GetDeclaringMethod, ReflectClas
 }
 FCIMPLEND
 
-FCIMPL1(ReflectClassBaseObject*, RuntimeTypeHandle::GetDeclaringType, ReflectClassBaseObject *pTypeUNSAFE) {
-    CONTRACTL {
-        FCALL_CHECK;
-    }
-    CONTRACTL_END;
+extern "C" TADDR QCALLTYPE RuntimeTypeHandle_GetDeclaringTypeHandleForGenericVariable(EnregisteredTypeHandle pTypeHandle)
```
Comment (jkotas) [inline src/coreclr/vm/runtimehandles.cpp:984]:
```suggestion
extern "C" EnregisteredTypeHandle QCALLTYPE RuntimeTypeHandle_GetDeclaringTypeHandleForGenericVariable(EnregisteredTypeHandle pTypeHandle)
```
Should this return `EnregisteredTypeHandle`?

Comment (jkotas) [reply src/coreclr/vm/runtimehandles.cpp:984]: (same for `RuntimeTypeHandle::GetElementTypeHandleFromHandle`)

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 3]
```diff
@@ -928,7 +967,14 @@ internal static string ConstructInstantiation(IRuntimeMethodInfo method, TypeNam
         }
 
         [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern RuntimeType GetDeclaringType(RuntimeMethodHandleInternal method);
+        private static extern unsafe MethodTable* GetMethodTable(RuntimeMethodHandleInternal method);
```
Comment (am11) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:970]: Looks like this type has an unused property https://github.com/dotnet/runtime/blob/c5235a64cf7a8e9448a6fe02df30c7c79607cbdd/src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs#L785

also maybe better to have one m_handle accessor visible:
```diff
        internal IntPtr Value => m_handle;
...
-        internal IntPtr m_handle;
+        private IntPtr m_handle;
```

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs [hunk 4]
```diff
@@ -48,13 +49,8 @@ private static RuntimeType GetTypeFromHandleSlow(IntPtr handle)
...
 
-        private static RuntimeType? GetTypeFromHandle(IntPtr handle)
+        private static RuntimeType GetRuntimeTypeFromHandle(IntPtr handle)
         {
-            if (handle == IntPtr.Zero)
-            {
-                return null;
-            }
-
             return GetTypeFromHandleIfExists(handle) ?? GetTypeFromHandleSlow(handle);
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs:54]: Rename `GetTypeFromHandleIfExists` -> `GetRuntimeTypeFromHandleIfExists`, `GetTypeFromHandleSlow` -> `GetRuntimeTypeFromHandleSlow` to match `GetTypeFromHandle` -> `GetRuntimeTypeFromHandle` rename?

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by AaronRobinsonMSFT** [comment]
@jkotas Looks like removing that `IsArray` check from `GetDeclaringType()` is causing linker legs to fail. Any thoughts?

I'll investigate the x86 leg tomorrow.

**Comment by jkotas** [comment]
> I'll investigate the x86 leg tomorrow.

This is failing in all PRs: https://github.com/dotnet/runtime/issues/110127

**Comment by jkotas** [comment]
> Looks like removing that IsArray check from GetDeclaringType() is causing linker legs to fail. Any thoughts?

The failures are caused by the linker assuming that instances of `RuntimeType` and `RuntimeAssembly` types are never allocated. The linker is replacing checks like `if (assembly is RuntimeAssembly)` by `if (false)` that is causing inscrutable failures.

The `RuntimeTypeHandle.GetRuntimeTypeFromHandle` call under the `IsArray` check is what makes the linker think that `RuntimeType` may be allocated. When the call is deleted as side-effect of deleting the `IsArray` check, the things fall apart. Of course, depending on a random call in the middle of reflection stack from keeping `RuntimeType` type alive is super fragile.

The problem is caused by the fact that the runtime allocates `RuntimeType` instances without running constructor. It is ok since the constructor would be no-op anyway, but it is confusing the linker. I think the proper fix is to add an entry for RuntimeType and RuntimeAssembly constructors under `#if FOR_ILLINK` to corelib.h to signal to the linker that the VM creates instances of these types. Once you do that, the linker tests should be passing with the `IsArray` check deleted.

**Review by jkotas — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
/ba-g "Failure is https://github.com/dotnet/dnceng/issues/3879. Not waiting for timeout."

**Pull request merged**